src/Features/LanguageServer/Protocol/Features/EditAndContinue/EditAndContinueSessionState.cs

```diff
@@ -0,0 +1,29 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.EditAndContinue;
+
+/// <summary>
+/// Manages in-proc EnC and Hot Reload session state that other components need to keep track of.
+/// </summary>
+/// <remarks>
+/// Separated from the in-proc EnC language service to allow access from lower-layer components.
+/// </remarks>
+[Export(typeof(IEditAndContinueSessionTracker))]
+[Export(typeof(EditAndContinueSessionState))]
+[Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class EditAndContinueSessionState() : IEditAndContinueSessionTracker
+{
+    public bool IsSessionActive { get; set; }
+
+    public ImmutableArray<DiagnosticData> ApplyChangesDiagnostics { get; set; } = [];
```

Comment:
can you doc threading expectations.  how does something that is consuming this know when teh data has properly been set on it?

also, as this is just returning a bool/diagnostics, could these just be helpers on some other interface? 

I'm not blocking this... it just seems... odd.  Not really a pattern we see elsewhere.

Finally, consider having two apis here, one that is read-only, and one that is read-write.  it would be good to separate out concerns.

Reply:
We have two APIs: IEditAndContinueSessionTracker and EditAndContinueSessionState

Reply:
Ah, i see.  Yes, that works for me.  Thanks.